pootle/apps/pootle_store/diff.py

```diff
@@ -0,0 +1,222 @@
+        # will be exactly what is in the file
+        if self.file_revision >= self.db_revision:
+            return self.file_unit_ids
+
+        # These units are kept as they have been updated since file_revision
+        # but do not appear in the file
+        new_units = [u for u in self.updated_db_units
+                     if u not in self.file_unit_ids]
+
+        # These unit are either present in both or only in the file so are
+        # kept in the file order
+        new_units += [u for u in self.file_unit_ids
+                      if u not in self.obsoleted_db_units]
+
+        return new_units
+
+    @cached_property
+    def obsoleted_db_units(self):
+        return [unitid for unitid, unit in self.db_units.items()
+                if unit['state'] == OBSOLETE
+                and unit["revision"] > self.file_revision]
+
+    @cached_property
+    def opcodes(self):
+        sm = difflib.SequenceMatcher(None, self.active_units,
+                                     self.new_unit_list)
+        return sm.get_opcodes()
+
+    @cached_property
+    def updated_db_units(self):
+        return [unitid for unitid, unit in self.db_units.items()
+                if unit['revision'] > self.file_revision
+                and not unit["state"] == OBSOLETE]
+
+    def get_updated_dbids(self):
+        """Returns a set of unit DB ids to be updated.
+        :return: a set of unit DB ids to be updated.
+        """
+        update_dbids = set()
+
+        def _has_changed(uid):
+            """ check if the source and targets match
+            or are equal to empty string
+            """
+            file_unit = self.file_units[uid]
+            db_unit = self.db_units[uid]
+
+            db_source = multistring_to_python(db_unit["source_f"])
+            db_target = multistring_to_python(db_unit["target_f"])
+
+            source_same = ((file_unit["source"] == "" and db_source == "") or
+                           file_unit["source"] == db_source)
+            target_same = ((file_unit["target"] == "" and db_target == "") or
+                           file_unit["target"] == db_target)
+
+            if source_same and target_same:
+                return False
+            return True
+
+        for (tag, i1, i2, j1, j2) in self.opcodes:
+            if tag == 'equal':
+                dbids = set(self.db_units[uid]['id']
+                            for uid in self.active_units[i1:i2]
+                            if uid in self.file_unit_ids
+                            and _has_changed(uid))
+                update_dbids.update(dbids)
+        return update_dbids
+
+    def diff(self):
+        """Return a dictionary of change actions"""
+        _diff = {}
+        _diff["index"] = self.get_indexes_to_update()
+        _diff["obsolete"] = self.get_units_to_obsolete()
+        _diff["add"] = self.get_units_to_add()
+        _diff["update"] = self.get_units_to_update()
+        return _diff
```

Comment:
Since there's no use in the extra variable, I'd directly return the dictionary (aligns with the rest of the codebase):
```python
return {
  'index': self.get_indexes_to_update(),
  'obsolete': self.get_units_to_obsolete(),
  'add': self.get_units_to_add(),
  'update': self.get_units_to_update(),
}
```